src/Microsoft.Azure.SignalR.Common/Utilities/BackOffPolicy.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    internal class BackOffPolicy
+    {
+        private TaskCompletionSource<bool> _currentProbeTcs = null;
+        private int _currentRetryCount = 0;
+
+        /// <summary>
+        /// Provides a synchronized mechanism of calling probing funcs by multiple concurrent callers.
+        /// Each caller's probe func will be invoked exactly one time.
+        /// The probe call may get delayed depending on the result of previous probe calls.
+        /// The delay is controlled by getRetryDelay func and the number of consecutive failed probe calls.
+        /// </summary>
+        /// <param name="probe"> this func returns a task with boolean result indicating if the probe was successful</param>
+        /// <param name="getRetryDelay"> this func returns a TimeSpan delay for a given iteration number</param>
+        /// <returns> 
+        /// A task with its final state and result matching the state and result of the task returned by the probe func.
+        /// This task will completes after either of the following happens:
+        /// - probe's task changes state to task.IsCompletedSuccessfully == true and task.Result == true
+        /// - probe's task.Result == false / faulted / cancelled and the delay defined by getRetryDelay has passed
+        /// </returns>
+        public async Task<bool> CallProbeWithBackOffAsync(Func<Task<bool>> probe, Func<int, TimeSpan> getRetryDelay)
+        {
+            bool calledProbeOnce = false;
+            bool probeSuccess = false;
+            do
+            {
+                var myTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Why is this making a new tcs in a loop?

Reply:
Hmm... There is no reason indeed. I guess I missed it because the cost is very low comparing to the new connection attempt. In the best case it gets allocated only once per call.